Use `HF_HUB_OFFLINE` + fix has_file in offline mode

Wauplin opened the pull request:
Should fix https://github.com/huggingface/transformers/issues/30345.
Related to https://github.com/huggingface/transformers/pull/31010 as well, even though it makes sense to merge both PRs IMO.

This PR:
- delegates offline mode to `huggingface_hub` with `HF_HUB_OFFLINE` env variable instead of `TRANSFORMERS_OFFLINE`. This change is still backward-compatible meaning that users that had already set `TRANSFORMERS_OFFLINE` won't notice a difference. The goal is to have a single environment variable to disable network in all HF libraries.
- updates the docs/tests to showcase `HF_HUB_OFFLINE` env variable
- make sure to respect offline mode in `has_file` => it was not the case before, causing https://github.com/huggingface/transformers/issues/30345 and https://github.com/huggingface/transformers/pull/31004
- add `local_files_only` and `cache_dir` to `has_file()` helper. At the moment, if connection/network is disabled, an error is raised. But if an authentication/authorization problem occurs, it returns False. IMO this is not consistent and we should aim at having the same response if server can't be requested. Instead of returning `False`, I suggest we default to checking the cache directory in case the information exists there.
- adds a test to check `has_file` in offline mode

Please let me know what you think.


(Note: some tests are failing but I prefer to get theoretical approval before moving forward on this).

Wauplin committed:
Fix has_file in offline mode

Wauplin committed:
harmonize env variable for offline mode

Wauplin committed:
Switch to HF_HUB_OFFLINE

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31016). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts reviewed (commented):
Thanks for working on this! 

Overall I think it looks good - only thing is that I think we need a deprecation cycle for `TRANSFORMERS_OFFLINE`

Wauplin committed:
fix test

Wauplin commented on tests/utils/test_hub_utils.py:
```diff
@@ -99,9 +101,20 @@ def test_non_existence_is_cached(self):
+            hf_hub_download(TINY_BERT_PT_ONLY, WEIGHTS_NAME, cache_dir=tmp_dir)
+
+            # Cache dir + offline mode => return True
+            assert has_file(TINY_BERT_PT_ONLY, WEIGHTS_NAME, cache_dir=tmp_dir)
```
Yes! Fixed in https://github.com/huggingface/transformers/pull/31016/commits/eb5968931fe1b0fa09befb98dcb9fe474a8acc71.

amyeroberts commented:
> Agree but if we want to deprecate it, I think we should do it in huggingface_hub directly.

OK - sounds good to me! 

> In any case I don't think we'll remove it before transformers v5, right?

I don't know, tbh, probably not before v5 as it's a common argument but if we think that's going to be too far away we could always move if forward

marianpascalau commented:
> Should fix #30345. Related to #31010 as well, even though it makes sense to merge both PRs IMO.
> 
> This PR:
> 
> * delegates offline mode to `huggingface_hub` with `HF_HUB_OFFLINE` env variable instead of `TRANSFORMERS_OFFLINE`. This change is still backward-compatible meaning that users that had already set `TRANSFORMERS_OFFLINE` won't notice a difference. The goal is to have a single environment variable to disable network in all HF libraries.
> * updates the docs/tests to showcase `HF_HUB_OFFLINE` env variable
> * make sure to respect offline mode in `has_file` => it was not the case before, causing [Problem with pretrained Transformers in Offline mode  #30345](https://github.com/huggingface/transformers/issues/30345) and [Do not trigger autoconversion if local_files_only #31004](https://github.com/huggingface/transformers/pull/31004)
> * add `local_files_only` and `cache_dir` to `has_file()` helper. At the moment, if connection/network is disabled, an error is raised. But if an authentication/authorization problem occurs, it returns False. IMO this is not consistent and we should aim at having the same response if server can't be requested. Instead of returning `False`, I suggest we default to checking the cache directory in case the information exists there.
> * adds a test to check `has_file` in offline mode
> 
> Please let me know what you think.
> 
> (Note: some tests are failing but I prefer to get theoretical approval before moving forward on this).

I want to make a suggestion. In order to better follow the implementation in "huggingface_hub", instead of following the envVariable you better follow the boolean variable (with the same name) from withing huggingface_hub module:

```
from huggingface_hub.constants import HF_HUB_OFFLINE

if HF_HUB_OFFLINE:
    print("Now we are offline")

```
This will better follow the huggingface implementation when they decide to change something.

Wauplin commented:
Hi @marianpascalau, thanks for the suggestion! Yes, that's exactly what I'm doing [here](https://github.com/huggingface/transformers/pull/31016/files#diff-82b93b530be62e40679876a764438660dedcd9cc9e33c2374ed21b14ebef5dbaR80) to delegate the env variable handling to `huggingface_hub`.

ydshieh commented:
Hi @Wauplin Is this PR also trying to fix the currently failing tests seen in other PRs too?

> python3 -m pytest -v tests/utils/test_offline.py::OfflineTests::test_offline_mode

Wauplin commented:
Hi @ydshieh I think https://github.com/huggingface/transformers/pull/31010 should fix it (but maybe this one as well). Both are kinda related

ydshieh commented:
Thanks! I will check

LysandreJik reviewed (approved):
Thanks @Wauplin ! To ensure that the deprecation cycle is respected, could we keep a test with `TRANSFORMERS_OFFLINE` while ensuring that it throws a warning?

As @amyeroberts said, this was a very well advertised argument

Wauplin committed:
revert test_offline to test TRANSFORMERS_OFFLINE

Wauplin committed:
Add new offline test

Wauplin commented:
Yes @LysandreJik @amyeroberts completly right. I reverted the tests to use `TRANSFORMERS_OFFLINE` (then no breaking change compared to `main`). And I added one to test that `HF_HUB_OFFLINE` is respected as well.

amyeroberts reviewed (approved):
Thanks for handling this!

Wauplin committed:
tMerge branch 'main' into 30345-fix-has-file-once-and-for-all

Wauplin committed:
merge conflicts

Wauplin committed:
docs

Wauplin commented:
@amyeroberts Tests are passing now. Feel free to do a last review or merge when it's ready :)

311-code commented:
I would also like to see added to hub.py's `def send_telemetry` something like:

`    def redact_sensitive_info(value): # Redact sensitive information such as paths or tokens if isinstance(value, str) and (os.path.exists(value) or 'token' in value.lower()): return "[REDACTED]" return value`

because how it is now seems like a bit of a security issue to me with the:

```
        args_as_dict = {k: v for k, v in args.__dict__.items() if not k.startswith("_") and v is not None}
        if "model_name_or_path" in args_as_dict:
            model_name = args_as_dict["model_name_or_path"]
```

and        ```
 if "dataset_name" in args_as_dict:
            data["dataset_name"] = args_as_dict["dataset_name"]
```
I noticed there were some bert models in 2022 trained on imdb and really hope this is not collecting training data on celebrity strings... It only seems to filter _ and None.

Wauplin commented:
Hi @311-code, concerns are legit but could you open a new issue on Github for it? This PR is related to offline mode only (and just so you know, when offline mode is enabled, telemetry is automatically stopped)

311-code commented:
Yes i won't be in front of PC until late tonight but I will highlight things I noticed. I am not quite sure if the redact code will work or have side effects yet.
